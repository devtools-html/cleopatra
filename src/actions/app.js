/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { getSelectedTab, getDataSource } from '../selectors/url-state';
import { getTrackThreadHeights } from '../selectors/app';
import { sendAnalytics } from '../utils/analytics';
import type { ThreadIndex } from '../types/profile';
import type { CssPixels } from '../types/units';
import type { Action, ThunkAction } from '../types/store';
import type { TabSlug } from '../app-logic/tabs-handling';
import type { UrlState } from '../types/state';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export const reportTrackThreadHeight = (
  threadIndex: ThreadIndex,
  height: CssPixels
): ThunkAction<void> => (dispatch, getState) => {
  const trackThreadHeights = getTrackThreadHeights(getState());
  const previousHeight = trackThreadHeights[threadIndex];
  if (previousHeight !== height) {
    // Guard against unnecessary dispatches. This could happen frequently.
    dispatch({
      type: 'UPDATE_TRACK_THREAD_HEIGHT',
      height,
      threadIndex,
    });
  }
};

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}
