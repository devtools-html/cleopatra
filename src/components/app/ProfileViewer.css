@keyframes profileViewerFadeIn {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
}

.profileViewer {
  animation-duration: 200ms;
  animation-name: profileViewerFadeIn;
  background-color: #fff;
}

.profileViewerName {
  height: 24px;
  padding: 0 6px;
  border-right: 1px solid var(--grey-30);
  color: var(--grey-60);
  font-weight: bold;
  line-height: 24px;
}

.profileViewerZipButton {
  /* Position */
  width: 18px;
  height: 18px;

  /* Box */
  flex: none;
  padding: 0;
  margin: 3px 0 3px 3px;
  border: 1px solid var(--green-60);

  /* Other */
  background: var(--green-50) url(../../../res/img/svg/back-arrow.svg) center
    center no-repeat;
  border-radius: 3px;
  color: #000;
}

.profileViewerZipButton:hover {
  border-color: #000;
}

.profileViewerSplitter {
  /* Reset the position: absolute that is set by the splitter-layout class */
  position: unset;
}

.profileViewerSplitter .layout-pane:not(.layout-pane-primary) {
  display: flex;
  overflow: hidden;
  max-height: var(--profile-viewer-splitter-max-height);
  flex-direction: column;
}
