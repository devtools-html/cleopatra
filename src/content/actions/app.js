// @flow
import type { Action } from './types';

export function changeSelectedTab(selectedTab: string): Action {
  return {
    type: 'CHANGE_SELECTED_TAB',
    selectedTab,
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function changeTabOrder(tabOrder: number[]): Action {
  return {
    type: 'CHANGE_TAB_ORDER',
    tabOrder,
  };
}
