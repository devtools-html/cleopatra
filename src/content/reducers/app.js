/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { combineReducers } from 'redux';

import type { Action } from '../actions/types';
import type { State, AppState, AppViewState, Reducer } from './types';

function view(state: AppViewState = { phase: 'INITIALIZING' }, action: Action): AppViewState {
  switch (action.type) {
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_WEB':
      return {
        phase: 'INITIALIZING',
        additionalData: { attempt: action.error.attempt },
      };
    case 'ERROR_RECEIVING_PROFILE_FROM_FILE':
    case 'ERROR_RECEIVING_PROFILE_FROM_ADDON':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_WEB':
      return { phase: 'FATAL_ERROR', error: action.error };
    case 'ROUTE_NOT_FOUND':
      return { phase: 'ROUTE_NOT_FOUND' };
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_WEB':
    case 'RECEIVE_PROFILE_FROM_FILE':
      return { phase: 'PROFILE' };
    default:
      return state;
  }
}

function isURLSetupDone(state: boolean = false, action: Action) {
  switch (action.type) {
    case '@@urlenhancer/urlSetupDone':
      return true;
    default:
      return state;
  }
}
const appStateReducer: Reducer<AppState> = combineReducers({ view, isURLSetupDone });
export default appStateReducer;

export const getApp = (state: State): AppState => state.app;
export const getView = (state: State): AppViewState => getApp(state).view;
export const getIsURLSetupDone = (state: State): boolean => getApp(state).isURLSetupDone;
