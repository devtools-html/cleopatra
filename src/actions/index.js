/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
