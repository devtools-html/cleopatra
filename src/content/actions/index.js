// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
