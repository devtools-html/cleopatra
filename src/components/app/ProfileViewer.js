/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import explicitConnect from '../../utils/connect';
import DetailsContainer from './DetailsContainer';
import ProfileFilterNavigator from './ProfileFilterNavigator';
import MenuButtons from './MenuButtons';
import SymbolicationStatusOverlay from './SymbolicationStatusOverlay';
import { returnToZipFileList } from '../../actions/zipped-profiles';
import { getProfileName } from '../../selectors/url-state';
import Timeline from '../timeline';
import { getHasZipFile } from '../../selectors/zipped-profiles';
import SplitterLayout from 'react-splitter-layout';
import { invalidatePanelLayout } from '../../actions/app';
import { getTimelineHeight } from '../../selectors/app';

import type { CssPixels } from '../../types/units';
import type { ConnectedProps } from '../../utils/connect';

require('./ProfileViewer.css');

type StateProps = {|
  +profileName: string | null,
  +hasZipFile: boolean,
  +timelineHeight: CssPixels | null,
|};

type DispatchProps = {|
  +returnToZipFileList: typeof returnToZipFileList,
  +invalidatePanelLayout: typeof invalidatePanelLayout,
|};

type Props = ConnectedProps<{||}, StateProps, DispatchProps>;

class ProfileViewer extends PureComponent<Props> {
  render() {
    const {
      hasZipFile,
      profileName,
      returnToZipFileList,
      invalidatePanelLayout,
      timelineHeight,
    } = this.props;

    return (
      <div
        className="profileViewer"
        style={
          timelineHeight === null
            ? {}
            : {
                '--profile-viewer-splitter-max-height': `${timelineHeight}px`,
              }
        }
      >
        <div className="profileViewerTopBar">
          {hasZipFile ? (
            <button
              type="button"
              className="profileViewerZipButton"
              title="View all files in the zip file"
              onClick={returnToZipFileList}
            />
          ) : null}
          {profileName ? (
            <div className="profileViewerName">{profileName}</div>
          ) : null}
          <ProfileFilterNavigator />
          {
            // Define a spacer in the middle that will shrink based on the availability
            // of space in the top bar. It will shrink away before any of the items
            // with actual content in them do.
          }
          <div className="profileViewerSpacer" />
          <MenuButtons />
        </div>
        <SplitterLayout
          customClassName="profileViewerSplitter"
          vertical
          percentage={false}
          // The DetailsContainer is primary.
          primaryIndex={1}
          // The Timeline is secondary.
          secondaryInitialSize={270}
          onDragEnd={invalidatePanelLayout}
        >
          <Timeline />
          <DetailsContainer />
        </SplitterLayout>
        <SymbolicationStatusOverlay />
      </div>
    );
  }
}

export default explicitConnect<{||}, StateProps, DispatchProps>({
  mapStateToProps: state => ({
    profileName: getProfileName(state),
    hasZipFile: getHasZipFile(state),
    timelineHeight: getTimelineHeight(state),
  }),
  mapDispatchToProps: {
    returnToZipFileList,
    invalidatePanelLayout,
  },
  component: ProfileViewer,
});
