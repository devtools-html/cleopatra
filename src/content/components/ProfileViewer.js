import React, { PureComponent, PropTypes } from 'react';
import { connect } from 'react-redux';
import TabBar from '../components/TabBar';
import ProfileSummaryView from '../containers/ProfileSummaryView';
import ProfileCallTreeView from '../containers/ProfileCallTreeView';
import ProfileMarkersView from '../containers/ProfileMarkersView';
import ProfileTaskTracerView from '../containers/ProfileTaskTracerView';
import ProfileLogView from '../containers/ProfileLogView';
import ProfileFilterNavigator from '../containers/ProfileFilterNavigator';
import ProfileSharing from '../containers/ProfileSharing';
import SymbolicationStatusOverlay from '../containers/SymbolicationStatusOverlay';
import TimelineView from '../containers/TimelineView';
import actions from '../actions';
import { getProfileViewOptions, getDisplayRange } from '../reducers/profile-view';
import { getSelectedTab } from '../reducers/url-state';
import ProfileViewerHeader from '../containers/ProfileViewerHeader';

class ProfileViewer extends PureComponent {
  constructor(props) {
    super(props);
    this._onSelectTab = this._onSelectTab.bind(this);

    // If updating this list, make sure and update the tabOrder reducer with another index.
    this._tabs = [
      {
        name: 'summary',
        title: 'Summary',
      },
      {
        name: 'calltree',
        title: 'Call Tree',
      },
      {
        name: 'markers',
        title: 'Markers',
      },
      {
        name: 'tasktracer',
        title: 'Task Tracer',
      },
      {
        name: 'log',
        title: 'Log',
      },
      {
        name: 'timeline',
        title: 'Timeline',
      },
    ];
  }

  _onSelectTab(selectedTab) {
    const { changeSelectedTab } = this.props;
    changeSelectedTab(selectedTab);
  }

  render() {
    const {
      className, viewOptions, timeRange, changeTabOrder, selectedTab,
    } = this.props;
    const { tabOrder } = viewOptions;
    return (
      <div className={className}>
        <div className={`${className}TopBar`}>
          <ProfileFilterNavigator />
          <ProfileSharing />
        </div>
        <ProfileViewerHeader />
        <TabBar tabs={this._tabs}
                selectedTabName={selectedTab}
                tabOrder={tabOrder}
                onSelectTab={this._onSelectTab}
                onChangeTabOrder={changeTabOrder} />
        {{
          summary: <ProfileSummaryView />,
          calltree: <ProfileCallTreeView />,
          markers: <ProfileMarkersView />,
          tasktracer: <ProfileTaskTracerView rangeStart={timeRange.start} rangeEnd={timeRange.end} />,
          timeline: <TimelineView />,
          log: <ProfileLogView />,
        }[selectedTab]}
        <SymbolicationStatusOverlay />
      </div>
    );
  }
}

ProfileViewer.propTypes = {
  className: PropTypes.string.isRequired,
  viewOptions: PropTypes.object.isRequired,
  timeRange: PropTypes.object.isRequired,
  selectedTab: PropTypes.string.isRequired,
  changeSelectedTab: PropTypes.func.isRequired,
  changeTabOrder: PropTypes.func.isRequired,
};

export default connect(state => ({
  viewOptions: getProfileViewOptions(state),
  selectedTab: getSelectedTab(state),
  className: 'profileViewer',
  timeRange: getDisplayRange(state),
}), actions)(ProfileViewer);
