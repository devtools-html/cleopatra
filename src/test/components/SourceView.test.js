/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React from 'react';
import { Provider } from 'react-redux';
import { stripIndent } from 'common-tags';

import { ProfileViewer } from 'firefox-profiler/components/app/ProfileViewer';
import { updateUrlState } from 'firefox-profiler/actions/app';
import { viewProfile } from 'firefox-profiler/actions/receive-profile';
import { stateFromLocation } from 'firefox-profiler/app-logic/url-handling';
import { ensureExists } from 'firefox-profiler/utils/flow';

import {
  render,
  screen,
  within,
} from 'firefox-profiler/test/fixtures/testing-library';
import { blankStore } from 'firefox-profiler/test/fixtures/stores';
import { getProfileFromTextSamples } from 'firefox-profiler/test/fixtures/profiles/processed-profile';
import { fireFullClick } from 'firefox-profiler/test/fixtures/utils';

// We're not interested in the timeline in this test
jest.mock('../../components/timeline', () => ({
  Timeline: 'custom-timeline',
}));

describe('SourceView', () => {
  function setup() {
    const revision = '997f00815e6bc28806b75448c8829f0259d2cb28';
    const filepath = 'widget/cocoa/nsAppShell.mm';
    window.fetch.get(
      `https://hg.mozilla.org/mozilla-central/raw-file/${revision}/${filepath}`,
      stripIndent`
        line 1
        line 2
        line 3
        line 4
        line 5
        line 6
        line 7
      `
    );

    const { profile } = getProfileFromTextSamples(`
      A[file:hg:hg.mozilla.org/mozilla-central:${filepath}:${revision}][line:4]
      B[file:git:github.com/rust-lang/rust:library/std/src/sys/unix/thread.rs:53cb7b09b00cbea8754ffb78e7e3cb521cb8af4b]
      C[lib:libC.so][file:s3:gecko-generated-sources:a5d3747707d6877b0e5cb0a364e3cb9fea8aa4feb6ead138952c2ba46d41045297286385f0e0470146f49403e46bd266e654dfca986de48c230f3a71c2aafed4/ipc/ipdl/PBackgroundChild.cpp:]
      D[lib:libD.so]
    `);

    const store = blankStore();
    store.dispatch(
      updateUrlState(
        stateFromLocation({
          pathname: '/from-browser',
          search: '',
          hash: '',
        })
      )
    );
    store.dispatch(viewProfile(profile));
    render(
      <Provider store={store}>
        <ProfileViewer />
      </Provider>
    );

    return {
      sourceView: () => document.querySelector('.sourceView'),
    };
  }

  it('does not show the source view at loadtime', () => {
    const { sourceView } = setup();
    expect(sourceView()).not.toBeInTheDocument();
  });

  it('should show the source view when double clicking on a line in the tree view', async () => {
    const { sourceView } = setup();

    const frameElement = screen.getByRole('treeitem', { name: /^A/ });
    fireFullClick(frameElement);
    fireFullClick(frameElement, { detail: 2 });
    expect(sourceView()).toBeInTheDocument();

    const sourceViewElement = ensureExists(sourceView());
    const sourceViewContent = await within(sourceViewElement).findByRole(
      'textbox'
    );

    // Because numbers and strings are split in several element, we're matching
    // on the string "line" only.
    await within(sourceViewContent).findAllByText('line');

    expect(sourceViewContent).toMatchSnapshot();
  });
});
