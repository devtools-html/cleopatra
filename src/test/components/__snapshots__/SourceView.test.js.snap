// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SourceView should show the source view when double clicking on a line in the tree view 1`] = `
<div
  aria-multiline="true"
  aria-readonly="true"
  autocapitalize="off"
  autocorrect="off"
  class="cm-content"
  contenteditable="false"
  role="textbox"
  spellcheck="false"
  style="tab-size: 4;"
  translate="no"
>
  <div
    class="cm-line"
  >
    <span
      class="cmt-typeName"
    >
      line 
    </span>
    <span
      class="cmt-number"
    >
      1
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="cmt-variableName"
    >
      line
    </span>
     
    <span
      class="cmt-number"
    >
      2
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="cmt-typeName"
    >
      line 
    </span>
    <span
      class="cmt-number"
    >
      3
    </span>
  </div>
  <div
    class="cm-nonZeroLine cm-line"
  >
    <span
      class="cmt-variableName"
    >
      line
    </span>
     
    <span
      class="cmt-number"
    >
      4
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="cmt-typeName"
    >
      line 
    </span>
    <span
      class="cmt-number"
    >
      5
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="cmt-variableName"
    >
      line
    </span>
     
    <span
      class="cmt-number"
    >
      6
    </span>
  </div>
  <div
    class="cm-line"
  >
    <span
      class="cmt-variableName"
    >
      line 
    </span>
    <span
      class="cmt-number"
    >
      7
    </span>
  </div>
</div>
`;
